moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.moshi.internal.canonicalize
+import com.squareup.moshi.internal.hasNullable
+import com.squareup.moshi.internal.jsonAnnotations
+import com.squareup.moshi.internal.toStringWithAnnotations
+import java.io.IOException
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Method
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+internal class AdapterMethodsFactory(
+  private val toAdapters: List<AdapterMethod>,
+  private val fromAdapters: List<AdapterMethod>
+) : JsonAdapter.Factory {
+  override fun create(
+    type: Type,
+    annotations: Set<Annotation>,
+    moshi: Moshi
+  ): JsonAdapter<*>? {
+    val toAdapter = get(toAdapters, type, annotations)
+    val fromAdapter = get(fromAdapters, type, annotations)
+    if (toAdapter == null && fromAdapter == null) return null
+
+    val delegate: JsonAdapter<Any>? = if (toAdapter == null || fromAdapter == null) {
+      try {
+        moshi.nextAdapter(this, type, annotations)
+      } catch (e: IllegalArgumentException) {
+        val missingAnnotation = if (toAdapter == null) "@ToJson" else "@FromJson"
+        throw IllegalArgumentException(
+          "No $missingAnnotation adapter for ${type.toStringWithAnnotations(annotations)}",
+          e
+        )
+      }
+    } else {
+      null
+    }
+
+    toAdapter?.bind(moshi, this)
+    fromAdapter?.bind(moshi, this)
+
+    return object : JsonAdapter<Any>() {
+      override fun toJson(writer: JsonWriter, value: Any?) {
+        when {
+          toAdapter == null -> {
+            requireNotNull(delegate).toJson(writer, value)
+          }
+          !toAdapter.nullable && value == null -> {
+            writer.nullValue()
+          }
+          else -> {
+            try {
+              toAdapter.toJson(moshi, writer, value)
+            } catch (e: InvocationTargetException) {
+              val cause = e.cause
+              if (cause is IOException) throw cause
+              throw JsonDataException("$cause at ${writer.path}", cause)
+            }
+          }
+        }
+      }
+
+      override fun fromJson(reader: JsonReader): Any? {
+        return when {
+          fromAdapter == null -> {
+            requireNotNull(delegate).fromJson(reader)
+          }
```

Comment:
```suggestion
          fromAdapter == null -> knownNotNull(delegate).fromJson(reader)
```

Same here re: known non null